core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -99,11 +102,15 @@ private[spark] class HeartbeatReceiver(sc: SparkContext)
   override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
     case heartbeat @ Heartbeat(executorId, taskMetrics, blockManagerId) =>
       if (scheduler != null) {
-        val unknownExecutor = !scheduler.executorHeartbeatReceived(
-          executorId, taskMetrics, blockManagerId)
-        val response = HeartbeatResponse(reregisterBlockManager = unknownExecutor)
         executorLastSeen(executorId) = System.currentTimeMillis()
-        context.reply(response)
+        reportExecutorHeartbeatReceived.submit(new Runnable {
```

Comment:
I think we can combine  `timeoutCheckingThread` and `reportExecutorHeartbeatReceived` to a single thread since the actions running in these threads are pretty fast.